src/JuliusSweetland.OptiKey/Models/DynamicKeyboardFolder.cs

```diff
@@ -0,0 +1,121 @@
+﻿using log4net;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Xml.Serialization;
+using JuliusSweetland.OptiKey.Extensions;
+
+namespace JuliusSweetland.OptiKey.Models
+{
+    struct KeyboardInfo
+    {
+        public string fullPath;
+        public string keyboardName;
+        public string symbolString;
+        public bool isHidden; // default false
+
+        public override string ToString()
+        {
+            string str = fullPath +  ", ";
+            str += keyboardName + ", ";
+            str += symbolString + ", ";
+            str += isHidden + ", ";
+            return str;
+        }
+    }
+
+    class DynamicKeyboardFolder
+    {
+
+        #region Constants
+
+        private const string ApplicationDataSubPath = @"JuliusSweetland\OptiKey\Keyboards\";
+        private const string OriginalKeyboardsSubPath = @"Keyboards\";
+
+        #endregion
+
+        #region Private Members
+
+        private static readonly ILog Log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        #endregion
+
+        public List<KeyboardInfo> keyboards;
+
+        public DynamicKeyboardFolder()
+        {
+            // Find all possible xml files
+            string filePath = GetUserKeyboardFolder();
+            string[] fileArray = Directory.GetFiles(filePath, "*.xml");
+
+            Log.InfoFormat("Found {0} keyboard files", fileArray.Length);
+
+            // Read in keyboard name, symbol, hidden state from each file
+            // Note that ordering is currently undefined
+            keyboards = new List<KeyboardInfo>();
+            foreach (string fileName in fileArray)
+            {
+                string keyboardPath = Path.Combine(filePath, fileName);
+                KeyboardInfo info = GetKeyboardInfo(keyboardPath);
+                if (null != info.fullPath)
+                {
+                    if (!info.isHidden)
+                    {
+                        keyboards.Add(info);
+                        Log.InfoFormat("Found keyboard file: {0}", info.fullPath);
+                    }
+                    else
+                    {
+                        Log.InfoFormat("Ignoring keyboard file: {0}", info.fullPath);
+                    }
+                }
+            }
+        }
+
+        #region Private Methods
+
+        private static string GetUserKeyboardFolder()
+        {
+            var applicationDataPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ApplicationDataSubPath);
+            Directory.CreateDirectory(applicationDataPath); //Does nothing if already exists
+
+            Log.DebugFormat("GetUserKeyboardFolder: {0}", applicationDataPath);
+
+            return applicationDataPath;
+        }
+
+        // get name from XML if present
```

Comment:
Could you add a setting to the management console (under the Visuals tab) for the user to specify the directory. Default to NULL, but on startup set this to the ApplicationData...Keyboards folder.

Reply:
I've added this. Some comments:
- defaults to empty string, then on startup it replaces with ApplicationData/etc/etc folder
- in management console, user sees current location (as a string) and selects new location from folder dialog
- If the directory is empty or doesn't exist any more, when launching the dynamic keyboards menu, extra diagnostic information is displayed to user on the empty keyboard selection grid (i.e. this is where I'm looking for keyboards, you can change it in management console/visuals)